x/proof/keeper/proof_validation.go

```diff
@@ -175,10 +154,87 @@ func (k Keeper) EnsureValidProof(
 	); err != nil {
-		return types.ErrProofInvalidRelayDifficulty.Wrapf("failed to validate relay difficulty for service %s due to: %v", sessionHeader.ServiceId, err)
+		return nil, types.ErrProofInvalidRelayDifficulty.Wrapf("failed to validate relay difficulty for service %s due to: %v", sessionHeader.ServiceId, err)
 	}
 	logger.Debug("successfully validated relay mining difficulty")
 
+	// Retrieve the corresponding claim for the proof submitted
+	claim, err := k.queryAndValidateClaimForProof(ctx, sessionHeader, supplierOperatorAddr)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debug("successfully retrieved and validated claim")
+
+	return claim, nil
+}
+
+// EnsureValidProofSignaturesAndClosestPath ensures that the proof submitted by
+// the supplier has a valid relay request/response signatures and closest path
+// with respect to an onchain claim.
+//
+// This function should be called in the EndBlocker corresponding to the block height
+// at which the proof is submitted rather than during proof submission (i.e. SubmitProof).
+//
+// NOTE: A fully valid proof must pass both EnsureWellFormedProof and
+// EnsureValidProofSignaturesAndClosestPath.
+func (k Keeper) EnsureValidProofSignaturesAndClosestPath(
+	ctx context.Context,
+	proof *types.Proof,
+) error {
+	// Telemetry: measure execution time.
+	defer cosmostelemetry.MeasureSince(cosmostelemetry.Now(), telemetry.MetricNameKeys("proof", "validation")...)
+
+	logger := k.Logger().With("method", "EnsureValidProofSignaturesAndClosestPath")
+
+	// Retrieve the supplier operator's public key.
+	supplierOperatorAddr := proof.SupplierOperatorAddress
+	supplierOperatorPubKey, err := k.accountQuerier.GetPubKeyFromAddress(ctx, supplierOperatorAddr)
+	if err != nil {
+		return err
+	}
+
+	sessionHeader := proof.GetSessionHeader()
+
+	// Unmarshal the closest merkle proof from the message.
+	sparseCompactMerkleClosestProof := &smt.SparseCompactMerkleClosestProof{}
+	if err = sparseCompactMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof); err != nil {
+		return types.ErrProofInvalidProof.Wrapf(
+			"failed to unmarshal closest merkle proof: %s",
+			err,
+		)
+	}
+
+	// SparseCompactMerkeClosestProof does not implement GetValueHash, so we need to decompact it.
+	sparseMerkleClosestProof, err := smt.DecompactClosestProof(sparseCompactMerkleClosestProof, &protocol.SmtSpec)
+	if err != nil {
+		return types.ErrProofInvalidProof.Wrapf(
+			"failed to decompact closest merkle proof: %s",
+			err,
+		)
+	}
+
+	// Get the relay request and response from the proof.GetClosestMerkleProof.
+	relayBz := sparseMerkleClosestProof.GetValueHash(&protocol.SmtSpec)
+	relay := &servicetypes.Relay{}
+	if err = k.cdc.Unmarshal(relayBz, relay); err != nil {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"failed to unmarshal relay: %s",
+			err,
+		)
+	}
+
+	// Verify the relay request's signature.
+	if err = k.ringClient.VerifyRelayRequestSignature(ctx, relay.GetReq()); err != nil {
```

Comment:
Are we 100% sure the ringClient is up to date here?

Reply:
Yes, the info is embedded into the application record and the ring client processes it correctly.